### PR TITLE
add isotonic regression scaling to SVM

### PR DESCRIPTION
**Pull request by rayeaster**
add IsotonicRegressionScaling as a new probability calibration method
https://github.com/haifengl/smile/issues/354

http://fastml.com/classifier-calibration-with-platts-scaling-and-isotonic-regression/
https://en.wikipedia.org/wiki/Isotonic_regression#Algorithms

image

### CODE REVIEW

**Comment by rayeaster** [comment]
Note that usually it is recommended to use separate data set other than traing data set for probability calibration, i.e., probability calibration should be done on new data not used for model fitting. Typically,  N-fold cross-validation could be used to split train samples and validation samples for the calibration. The probabilities predicted for the folds are then averaged.

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            throw new IllegalArgumentException("fail to get the posteriori probability for given prediction: " + y);
+        }
+        return ret;
+    }
+
+    /**
+     * Estimates the multiclass probabilies.
+     */
+    public static void multiclass(int k, double[][] r, double[] p) {
+        PlattScaling.multiclass(k, r, p);
```
Comment (haifengl) [inline core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: Can we do this? Do you have a reference?

File: core/src/main/java/smile/classification/SVM.java [hunk 2]
```diff
@@ -1533,15 +1698,19 @@ private double posterior(LASVM svm, double y) {
         final double minProb = 1e-7;
         final double maxProb = 1 - minProb;
 
-        return Math.min(Math.max(svm.platt.predict(y), minProb), maxProb);
+        return Math.min(Math.max(scalingMethod == ScalingMethod.Platt? svm.platt.predict(y) : svm.isotonicRegression.predict(y), minProb), maxProb);
```
Comment (haifengl) [inline core/src/main/java/smile/classification/SVM.java]: no sure about this. simple prediction without postiori probability should not bother this at all.

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java (hunk 1 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: No. Reference for platt does not mention multi-class calculation neither: https://www.researchgate.net/profile/John_Platt/publication/2594015_Probabilistic_Outputs_for_Support_Vector_Machines_and_Comparisons_to_Regularized_Likelihood_Methods/links/004635154cff5262d6000000.pdf

File: core/src/main/java/smile/classification/SVM.java (hunk 2 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/SVM.java]: this is just a switch added on top of existing logic. If simple prediction without postiori probability does not need concern about scaling, why do we need existing platt scaling in this method?

**Comment by rayeaster** [comment]
add friendly `toString` to output pair-wise lines generated by Isotonic Regression for better illustration of the scaling effect. Also add straightforward assertion check to ensure we cover all prediction in fitted lines.

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    private double trainMaxScore = Double.NaN;
+
+    /** setep-wise constant functions */
+    private static class StepwiseConsFunc {
+        double lowScore;
+        double highScore;
+        double val;
+        int weight = 1;
+
+        StepwiseConsFunc next;
```
Comment (haifengl) [inline core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: Java has LinkedList. We don't need to reinvent the wheel. In fact, we don't need linked list at all.

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        StepwiseConsFunc previous;
+
+        StepwiseConsFunc(double lowScore, double highScore, double val) {
+            this.lowScore = lowScore;
+            this.highScore = highScore;
+            this.val = val;
+        }
+
+        void setNext(StepwiseConsFunc next) {
```
Comment (haifengl) [inline core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: It is not necessary to use getter and setter for inner class. They make codes cluttered and don't really bring in benefits of encapsulations. You are the only user of this class!

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+     * @param y
+     *            The training labels.
+     */
+    public IsotonicRegressionScaling(double[] scores, int[] y) {
+        initStepwiseConstFuncs(scores, y);
+
+        StepwiseConsFunc current = startConstantFuncs.next;
+        StepwiseConsFunc last = startConstantFuncs;
+        
+        while (current != null) {
```
Comment (haifengl) [inline core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: We need actually check the PAV in multiple iterations. Think about it as an exercise.

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+ * <li>Niculescu-Mizil, Alexandru; Caruana, Rich: Predicting Good Probabilities
+ * With Supervised Learning, 2005.</li>
+ * </ol>
+ *
+ * @author rayeaster
+ */
+public class IsotonicRegressionScaling implements Serializable {
+    private static final long serialVersionUID = 1L;
+    
+    private double trainMixScore = Double.NaN;
```
Comment (haifengl) [inline core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: They are working variables. Should not be member variables.

File: core/src/main/java/smile/classification/IsotonicRegressionScaling.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    /**
+     * Returns the posterior probability estimate P(y = 1 | x).
+     *
+     * @param y
+     *            the binary classifier output score.
+     * @return the estimated probability.
+     */
+    public double predict(double y) {
```
Comment (haifengl) [inline core/src/main/java/smile/classification/IsotonicRegressionScaling.java]: This is really slow. It is linear. It can be done in O(log n).

**Comment by haifengl** [comment]
I implement it in v2 now. Thanks!

**Comment by haifengl** [comment]
check out https://github.com/haifengl/smile/blob/v2/core/src/main/java/smile/classification/IsotonicRegressionScaling.java

**Comment by rayeaster** [comment]
I will close this PR

**Pull request closed without merging**